### PR TITLE
JDBC BigDecimal support

### PR DESCRIPTION
**Pull request by Jens-H**
This PR changes the way the DECIMAL sql type is matched to a Java type.

Problem:
The actual way is to always cast the decimal value to double. This seems like an unfortunate cast as the decimal data type tries to avoid some of the unwanted properties of floating point numbers.

Solution:
According to the JDBC reference the recommended Java data type would be BigDecimal. This PR changes/implements the calls to getObject/setObject with a BigDecimal object and getBigDecimal/setBigDecimal. 
To reflect the different sizes a sql decimal can have, it was necessary to introduce a new meta data class DuckDBColumnTypeMetaData that holds the additional information regarding scale and precision.

As always additional test are added.

Note: This PR builds up on not yet merged changes from #2970

### CODE REVIEW

File: tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java [hunk 1]
```diff
@@ -811,10 +1073,10 @@ public static void test_hugeint() throws Exception {
 		assertEquals(rs.getLong("hi2"), -42L);
 		assertEquals(rs.getObject("hi3"), new BigInteger("454564646545646546545646545"));
 		assertEquals(rs.getObject("hi4"), new BigInteger("-454564646545646546545646545"));
-		assertEquals(rs.getBigDecimal("hi1"), new BigDecimal("42"));
-		assertEquals(rs.getBigDecimal("hi2"), new BigDecimal("-42"));
-		assertEquals(rs.getBigDecimal("hi3"), new BigDecimal("454564646545646546545646545"));
-		assertEquals(rs.getBigDecimal("hi4"), new BigDecimal("-454564646545646546545646545"));
+		// assertEquals(rs.getBigDecimal("hi1"), new BigDecimal("42"));
```
Comment (Mytherin) [inline tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:1076]: Should this be commented out?

File: tools/jdbc/src/jni/duckdb_java.cpp [hunk 2]
```diff
@@ -226,6 +250,42 @@ JNIEXPORT void JNICALL Java_org_duckdb_DuckDBNative_duckdb_1jdbc_1free_1result(J
...
+		uint8_t scale = 0;
+		logical_type.GetDecimalProperties(width, scale);
+		std::string width_scale = std::to_string(width) + std::string(";") + std::to_string(scale);
+
+		auto physical_type = logical_type.InternalType();
+		switch (physical_type) {
+		case PhysicalType::INT16: {
+			string res = std::string("DECIMAL16;") + width_scale;
+			return res;
+		} break;
```
Comment (Mytherin) [inline tools/jdbc/src/jni/duckdb_java.cpp:267]: nit: the breaks here are not necessary

File: tools/jdbc/src/jni/duckdb_java.cpp [hunk 3]
```diff
@@ -322,9 +388,26 @@ JNIEXPORT jobjectArray JNICALL Java_org_duckdb_DuckDBNative_duckdb_1jdbc_1fetch(
...
+				break;
+			case PhysicalType::INT64:
+				constlen_data = env->NewDirectByteBuffer(FlatVector::GetData(vec), row_count * sizeof(int64_t));
+				break;
+			case PhysicalType::INT128:
+				constlen_data = env->NewDirectByteBuffer(FlatVector::GetData(vec), row_count * sizeof(hugeint_t));
+				break;
+			}
+			break;
+			/* Vector double_vec(LogicalType::DOUBLE); */
```
Comment (Mytherin) [inline tools/jdbc/src/jni/duckdb_java.cpp:408]: Perhaps remove the comments here?

File: tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java [hunk 4]
```diff
@@ -410,7 +412,207 @@ public static void duckdb_timestamp_test() throws Exception {
...
+		assertEquals(rs.getTimestamp(10), rs.getObject(10, Timestamp.class));
+		assertEquals(rs.getObject(10, LocalDateTime.class), LocalDateTime.parse("1970-01-03T03:42:23"));
+		assertEquals(rs.getObject(10, LocalDateTime.class), LocalDateTime.of(1970,1,3,3,42,23));
+		assertEquals(rs.getBigDecimal(11), rs.getObject(11, BigDecimal.class)); 
+		assertEquals(rs.getBigDecimal(12), rs.getObject(12, BigDecimal.class)); 
+		assertEquals(rs.getBigDecimal(13), rs.getObject(13, BigDecimal.class)); 
+		assertEquals(rs.getBigDecimal(14), rs.getObject(14, BigDecimal.class)); 
+
+		// Missing implementations, should never reach assertTrue(false)
+		try {
```
Comment (Mytherin) [inline tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:474]: Can we use `assertThrows` here?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Apologies for the delay in reviewing. Looks good to me. Some minor comments:

File: tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java (hunk 4 above)
Comment (Jens-H) [reply tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java:474]: Unfortunately there is no `assertThrows` method implemented. It is rather easy to do this with lambdas and have something like `assertThrows(() -> {return rs.getObject(11, Integer.class);});`, but these are not available in Java 7. 

Implementations without lambdas are certainly possible, but if then the code is less messy is another question.

File: tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java (hunk 1 above)
Comment (Jens-H) [reply tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java]: I reworked the handling in the case of calling getBigDecimal on a non-decimal result. Now the tests are OK again, but this is on best effort basis and might break for other numbers.

In addition the tests now use the compareTo method, which ignores differences in scale as long as the numbers are identical. As we are heavily casting here, this seems to be appropriate.

**Comment by Mytherin** [comment]
Thanks! Looks good.

**Pull request merged**